src/inference/AdvancedSMC.jl

```diff
@@ -58,35 +100,22 @@ function step(model, spl::Sampler{<:SMC}, vi::VarInfo)
     ## pick a particle to be retained.
     Ws, _ = weights(particles)
     indx = randcat(Ws)
-    push!(spl.info[:logevidence], particles.logE)
-
-    return particles[indx].vi, true
-end
+    push!(spl.state.logevidence, particles.logE)
 
-## wrapper for smc: run the sampler, collect results.
-function sample(model::Model, alg::SMC)
-    spl = Sampler(alg)
+    # update the master vi.
+    spl.state.vi = particles[indx].vi
+    params = tonamedtuple(spl.state.vi)
+    lp = getlogp(spl.state.vi)
 
-    particles = ParticleContainer{Trace}(model)
-    push!(particles, spl.alg.n_particles, spl, empty!(VarInfo(model)))
-
-    while consume(particles) != Val{:done}
-      ess = effectiveSampleSize(particles)
-      if ess <= spl.alg.resampler_threshold * length(particles)
-        resample!(particles,spl.alg.resampler)
-      end
-    end
-    w, samples = getsample(particles)
-    res = Chain(log(w), samples)
-    return res
+    return transition(params, lp, Ws[indx], particles.logE)
```

Comment:
Here all particles should be returned since `SMC` is not an iterative algorithm. That is, it can only be run forwardly once, then return all particles (in the form of a `MCMCChains.Chain`). This is different from `PG` or `ConditionalSMC`, which is an MH algorithm and can be run iteratively and each iteration produces one sample (i.e. transition).

Reply:
So in the case of `SMC` is a `step!` function even appropriate? How would this mix with Gibbs, which calls a `step!` for each of it's component samplers?

Reply:
For `IS`, `SMC` like samplers, `step!` can only be run for one iteration. We can't really use SMC as a component sampler for Gibbs since it's like a valid MH kernel. However, I think it's ok to keep the interface here and add a comment saying that it shouldn't be run iteratively.

Reply:
Ah, okay. So should each sample in SMC track the log evidence? If we only ever do one sweep then it's not really clear what the `logevidence` for each sample would be, since they'd all come from the same particle container.

Reply:
@cpfiffer @trappmartin  is this what you are looking for in another comment?

Reply:
Yeah, it came up, but I think we're good now?